### PR TITLE
VM handle large values when accessing byte arrays - remove optimistic…

### PR DESCRIPTION
**Pull request by rtkaczyk**
… toInt conversions

Fixes block 549413

### CODE REVIEW

**Comment by rtkaczyk** [comment]
Tests added

File: src/main/scala/io/iohk/ethereum/vm/OpCode.scala [hunk 1]
```diff
@@ -163,8 +163,11 @@ object OpCodes {
 }
 
 object OpCode {
-  def sliceBytes(bytes: ByteString, offset: Int, size: Int): ByteString =
-    bytes.slice(offset, offset + size).padTo(size, 0.toByte)
+  def sliceBytes(bytes: ByteString, offset: UInt256, size: UInt256): ByteString = {
+    val start = offset.min(bytes.size).toInt
+    val end = (offset + size).min(bytes.size).toInt
+    bytes.slice(start, end).padTo(size.toInt, 0.toByte)
```
Comment (rtkaczyk) [inline src/main/scala/io/iohk/ethereum/vm/OpCode.scala]: I think `.padTo(end - start, 0.toByte)` would make more sense. But this is what [geth does](https://github.com/ethereum/go-ethereum/blob/master/core/vm/common.go#L42). See the consequence in the tests.

Comment (adamsmo) [inline src/main/scala/io/iohk/ethereum/vm/OpCode.scala]: maybe but for me it is more readable this way

File: src/main/scala/io/iohk/ethereum/vm/OpCode.scala [hunk 2]
```diff
@@ -367,7 +370,7 @@ case object CODESIZE extends ConstOp(0x38)(_.env.program.length)
 case object CODECOPY extends OpCode(0x39, 3, 0, _.G_verylow) {
   protected def exec[W <: WorldStateProxy[W, S], S <: Storage[S]](state: ProgramState[W, S]): ProgramState[W, S] = {
     val (Seq(memOffset, codeOffset, size), stack1) = state.stack.pop(3)
-    val bytes = state.program.getBytes(codeOffset.toInt, size.toInt)
+    val bytes = OpCode.sliceBytes(state.program.code, codeOffset, size)
```
Comment (LukasGasior1) [inline src/main/scala/io/iohk/ethereum/vm/OpCode.scala]: Do we still need `state.program.getBytes`?

Comment (rtkaczyk) [inline src/main/scala/io/iohk/ethereum/vm/OpCode.scala]: Yes, it's used for PUSHx (the semantics would be the same, but no unnecessary UInt256 -> Int conversions)

File: src/main/scala/io/iohk/ethereum/vm/OpCode.scala (hunk 1 above)
Comment (rtkaczyk) [inline src/main/scala/io/iohk/ethereum/vm/OpCode.scala]: It's not about readability...

Comment (rtkaczyk) [inline src/main/scala/io/iohk/ethereum/vm/OpCode.scala]: Compare the 5th and 6th table rows (counting from 1)

Comment (adamsmo) [inline src/main/scala/io/iohk/ethereum/vm/OpCode.scala]: 👍

**Pull request merged**